magicblock-processor/src/scheduler/coordinator.rs

```diff
@@ -0,0 +1,239 @@
+        if let Err(index) = index {
+            queue.insert(index, transaction);
+        }
+        executor
+    }
+
+    /// Checks if there are any executors ready to process a transaction.
+    #[inline]
+    pub(super) fn is_ready(&self) -> bool {
+        !self.ready_executors.is_empty()
+    }
+
+    /// Retrieves the ID of a ready executor, if one is available.
+    #[inline]
+    pub(super) fn get_ready_executor(&mut self) -> Option<ExecutorId> {
+        self.ready_executors.pop()
+    }
+
+    /// Returns an executor to the pool of ready executors.
+    #[inline]
+    pub(super) fn release_executor(&mut self, executor: ExecutorId) {
+        self.ready_executors.push(executor)
+    }
+
+    /// Releases all account locks held by a specific executor.
+    pub(crate) fn unlock_accounts(&mut self, executor: ExecutorId) {
+        let locks = &mut self.acquired_locks[executor as usize];
+        // Iteratively drain the list of acquired locks.
+        while let Some(lock) = locks.pop() {
+            lock.borrow_mut().unlock(executor);
+        }
+    }
+
+    /// Retrieves the next blocked transaction waiting for a given executor.
+    pub(super) fn next_blocked_transaction(
+        &mut self,
+        executor: ExecutorId,
+    ) -> Option<TransactionWithId> {
+        self.blocked_transactions[executor as usize].pop_front()
+    }
+
+    /// Attempts to acquire all necessary read and write locks for a transaction.
+    ///
+    /// This function iterates through all accounts in the transaction's message and
+    /// attempts to acquire the appropriate lock for each. If any lock is contended,
+    /// it fails early and returns the ID of the blocking executor or transaction.
+    pub(super) fn try_acquire_locks(
+        &mut self,
+        executor: ExecutorId,
+        transaction: &TransactionWithId,
+    ) -> Result<(), TransactionId> {
+        let message = transaction.txn.transaction.message();
+        let accounts_to_lock = message.account_keys().iter().enumerate();
+        let acquired_locks = &mut self.acquired_locks[executor as usize];
+
+        for (i, &acc) in accounts_to_lock.clone() {
+            // Get or create the lock for the account.
+            let lock = self.locks.entry(acc).or_default().clone();
+            // See whether there's a contention for the given account
+            // if there's one, then we need to follow the breadcrumbs
+            // (txns->executor) to find out where our transaction
+            // needs to be queued.
+            let mut result =
+                if let Some(contenders) = self.account_contention.get(&acc) {
+                    match contenders.binary_search(&transaction.id) {
+                        // If we are the first contender, then we can proceed
+                        // and try to acquire the needed account locks
+                        Ok(index) | Err(index) if index == 0 => Ok(()),
```

Comment:
```suggestion
                        Ok(0) | Err(0) => Ok(()),
```

Correct me if I'm wrong
Ok(0) - means that transaction is being rescheduled
Err(0) - new transaction.

If this is correct - would be good to add this as a doc